_posts/2015/2015-06-29-git-migrate.markdown

```diff
@@ -0,0 +1,153 @@
+
+### The Easy Case - Fixing `master`
+
+This is the simple case. I made commits on master that were intended for a branch off of master. Lets walk through this scenario step by step with some visual aids.
+
+The following diagram shows the state of my repository before I got all itchy trigger finger on it.
+
+![Initial state](https://cloud.githubusercontent.com/assets/19977/8390497/bbde5f8e-1c4c-11e5-9760-9e94236423d6.png)
+
+As you can see, I have two commits to the `master` branch. HEAD points to the tip of my current branch. You can also see a remote tracking branch named `origin/master` (_this is a special branch that tracks the `master` branch on the remote server_). So at this point, my local `master` matches the `master` on the server.
+
+This is the state of my repository when I am struck by inspiration and I start to code.
+
+![First](https://cloud.githubusercontent.com/assets/19977/8390499/bbe3174a-1c4c-11e5-87ee-75aacf1a197a.png)
+
+I make one commit. Then two.
+
+![Second Commit - fixing time](https://cloud.githubusercontent.com/assets/19977/8390498/bbe2c524-1c4c-11e5-98c3-7526ae2277f9.png)
+
+Each time I make a commit, the local `master` branch is updated to the new commit. Uh oh! As in the scenario in the opening paragraph, I meant to create these two commits on a new branch creatively named `new-branch`. I better fix this up.
+
+So the first thing I do is create this branch without switching to it.
+
+`git branch new-branch`
+
+![Create a new branch named new-branch](https://cloud.githubusercontent.com/assets/19977/8390500/bbe365ec-1c4c-11e5-83e6-eb60f7155efe.png)
+
+Notice that `HEAD` still points to `master`. Also note that `new-branch` points to the same location. This is why branching is so cheap. A branch is just a pointer to a commit.
+
+I created a branch that points to the latest commit to make it safe for me to reset the master branch back to the state it is on the server. I want to make sure I have an easy way to still reference the commit that the branch points to before I reset it.
+
+`git reset origin/master --hard`
+
+![Reset master back to its original position](https://cloud.githubusercontent.com/assets/19977/8390501/bbf39b24-1c4c-11e5-8d7a-b1cf7b4bcb5b.png)
+
+Notice that `HEAD` and `master` are back to their original position, but `new-branch` continues to point to the latest commit.
+
+Now I can simply check out the `new-branch` and continue my work.
+
+![Checkout new-branch and we are done](https://cloud.githubusercontent.com/assets/19977/8390502/bbf69180-1c4c-11e5-920a-7991e341eefa.png)
+
+This moves `HEAD` to the new branch and now I can continue to work on the branch as if I had never made this mistake. Yay!
+
+Here's the set of commands that I ran all together.
+
+```bash
+git branch new-branch
+git reset origin/master --hard
+git checkout new-branch
+```
+
```

Comment:
It is true that the most straightforward approach is these three commands. But it is awkward because it involves rewriting the working tree twice (as you jump back to `origin/master` then return to `new-branch`). The most serious problem with this is that any uncommitted changes to existing files will be lost.

In general, `git reset --hard` is dangerous. It actively and silently discards uncommitted work. Sometimes that's what you want, but (I think) not in this case. So in this case `git reset --keep` would be preferable: it does the same thing as `git reset --hard` but (1) tries to carry any uncommitted changes over, but (2) if it can't it reports an error and aborts.

But this can be reduced to two commands that *don't* involve rewriting the working tree:

```
git checkout -b new-branch
git branch --force master origin/master
```

I wasn't sure about it but I just verified that the second command retains the reflog for `master` (including the addition of the unwanted commits and the reset back to `origin/master`), which is a good thing.

You can even do this in one command, kindof:

```
git branch -m new-branch
```

This renames your local `master` branch to `new-branch`. It has two strong caveats:

* The reflog for `master` is also moved to `new-branch`. If you don't want to lose the backstory for `master`, this is not so good.
* This command leaves the repo without a `master` branch. This is not as bad as it sounds, because whenever you want to go back to work you can just type `git checkout master` and the branch will be re-created automatically from `origin/master`. But if your local `master` had unpushed commits that you didn't want to be moved to `new-branch`, you can't use this version.